server-utils/src/suspendable_stream.rs

```diff
@@ -40,30 +40,34 @@ impl<S, I> Stream for SuspendableStream<S>
 	type Error = ();
 
 	fn poll(&mut self) -> Result<Async<Option<Self::Item>>, ()> {
-		if let Some(mut timeout) = self.timeout.take() {
-			match timeout.poll() {
-				Ok(Async::Ready(_)) => {}
-				Ok(Async::NotReady) => {
-					self.timeout = Some(timeout);
-					return Ok(Async::NotReady);
+		loop {
+			if let Some(mut timeout) = self.timeout.take() {
+				match timeout.poll() {
+					Ok(Async::Ready(_)) => {}
+					Ok(Async::NotReady) => {
+						self.timeout = Some(timeout);
+						return Ok(Async::NotReady);
+					}
+					Err(err) => {
+						warn!("Timeout error {:?}", err);
+						task::current().notify();
```

Comment:
Sorry to chime in so late, but this doesn't make any sense at all? Why notify the task instead of just jumping to `self.stream.poll()`? There is nothing else that may happen when you `notify` the task and return `NotRead` right away.

Reply:
If we have `self.timeout.is_some() == true`, we polled the `Stream` and it did return an error, that does indicate that we can not accept any more connections at the moment. So, we created a timeout and we are at polling this timeout and the timeout returns an error. This error is very unlikely, but could happen if we already have a lot of other timers registered in the Tokio runtime. 
So, we can not poll `self.stream` as we can not accept any new connections and polling the `timeout` also returned an error.
However, we want that other futures can continue and maybe finish. For that, we inform the current task, that our future should be polled in the next cycle again and return `NotReady`.
Now, Tokio will continue with all the other futures that are currently ready and need to be processed and in the next "cycle" our future is called again.

Reply:
Ok, thanks for the explanation. Although `self.timeout.is_some()` doesn't imply that `stream` was polled in the same cycle, but indeed the behavior here is correct and handles all the cases, my proposed change would probably have this subtle bug, when both stream and timeout errors and we get into this infinite loop of erroring.